cpp/src/io/fst/agent_dfa.cuh

```diff
@@ -0,0 +1,675 @@
+    }
+
+    callback_op.TearDown();
+  }
+};
+
+template <bool IS_TRANS_VECTOR_PASS,
+          bool IS_SINGLE_PASS,
+          typename DfaT,
+          typename TileStateT,
+          typename AgentDFAPolicy,
+          typename SymbolItT,
+          typename OffsetT,
+          typename StateVectorT,
+          typename OutOffsetScanTileState,
+          typename TransducedOutItT,
+          typename TransducedIndexOutItT,
+          typename TransducedCountOutItT>
+__launch_bounds__(int32_t(AgentDFAPolicy::BLOCK_THREADS)) __global__
+  void SimulateDFAKernel(DfaT dfa,
+                         SymbolItT d_chars,
+                         OffsetT const num_chars,
+                         StateIndexT seed_state,
+                         StateVectorT* __restrict__ d_thread_state_transition,
+                         TileStateT tile_state,
+                         OutOffsetScanTileState offset_tile_state,
+                         TransducedOutItT transduced_out_it,
+                         TransducedIndexOutItT transduced_out_idx_it,
+                         TransducedCountOutItT d_num_transduced_out_it)
+{
+  using AgentDfaSimT = AgentDFA<AgentDFAPolicy, SymbolItT, OffsetT>;
+
+  static constexpr int32_t NUM_STATES = DfaT::MAX_NUM_STATES;
+
+  enum {
+    BLOCK_THREADS     = AgentDFAPolicy::BLOCK_THREADS,
+    ITEMS_PER_THREAD  = AgentDFAPolicy::ITEMS_PER_THREAD,
+    SYMBOLS_PER_BLOCK = AgentDfaSimT::SYMBOLS_PER_BLOCK
+  };
+
+  // Shared memory required by the DFA simulation algorithm
+  __shared__ typename AgentDfaSimT::TempStorage dfa_storage;
+
+  // Shared memory required by the symbol group lookup table
+  __shared__ typename DfaT::SymbolGroupStorageT symbol_matcher_storage;
+
+  // Shared memory required by the transition table
+  __shared__ typename DfaT::TransitionTableStorageT transition_table_storage;
+
+  // Shared memory required by the transducer table
+  __shared__ typename DfaT::TranslationTableStorageT transducer_table_storage;
+
+  // Initialize symbol group lookup table
+  auto symbol_matcher = dfa.InitSymbolGroupLUT(symbol_matcher_storage);
+
+  // Initialize transition table
+  auto transition_table = dfa.InitTransitionTable(transition_table_storage);
+
+  // Initialize transition table
+  auto transducer_table = dfa.InitTranslationTable(transducer_table_storage);
+
+  // Set up DFA
+  AgentDfaSimT agent_dfa(dfa_storage);
+
+  // The state transition vector passed on to the second stage of the algorithm
+  StateVectorT out_state_vector;
+
+  // Stage 1: Compute the state-transition vector
+  if (IS_TRANS_VECTOR_PASS || IS_SINGLE_PASS) {
+    // Keeping track of the state for each of the <NUM_STATES> state machines
+    std::array<StateIndexT, NUM_STATES> state_vector;
+
+    // Initialize the seed state transition vector with the identity vector
+#pragma unroll
+    for (int32_t i = 0; i < NUM_STATES; ++i) {
+      state_vector[i] = i;
+    }
```

Comment:
I guess we can't use `thrust` in this file?

Reply:
https://thrust.github.io/doc/group__transformations_ga9d17b70466df6b9a57f94170935a03f8.html
I don't know if it might be slower than the unrolled loop.

Reply:
Yup, algorithms over loops. Adopted your suggestion 👍